### ISSUE TITLE
Investigate how the gap between local and Dask predictions can be decreased

### ISSUE THREAD

**Issue by StrikerRUS** [microsoft/LightGBM#3835]
## Summary

Right now in our CI tests difference between local and Dask-based estimators are quite big in terms of produced results after training with the same set of params. It would be good to investigate reasons and either provide source code fixes or write some guide how the difference can be mitigated on a user side.

## References

https://github.com/microsoft/LightGBM/pull/3515#discussion_r541801734
- https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L161
- https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L206-L207
- https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L304-L306
- https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L436-L440
- ...

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by jameslamb** [comment]
I'd like to add that for this issue, we also need to investigate the gap for training with the same parameters and data using the non-Dask distributed training via the LightGBM CLI (https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#run-parallel-learning). That will tell us which differences are due to the way the Dask module works, and which are features  of LightGBM itself.

The same is true for networking issues.

At some point, I'm planning to add support for that in https://github.com/jameslamb/lightgbm-dask-testing so it's easy for anyone to test.

**Comment by shiyu1994** [comment]
@jmoralez Sorry for the late response. 

I suppose so IF the data partition are not identical in the CLI and Dask versions. Otherwise, I need more efforts to investigate where the difference came from.

In both CLI and Dask version of the above example, they should both use `pre_partition=True`, which means each process holds only part of the data.

For example, in the CLI case, each process will have `1000/2 = 500` instances. Each process will be responsible for finding the bin boundaries for a part of the features, with ONLY LOCAL training data, this is verified when I investigate the size of `sample_values[start[rank] + i]` here
https://github.com/microsoft/LightGBM/blob/2394b41ac2bc64c36115c4eb314633dede0c6b9c/src/io/dataset_loader.cpp#L716-L719

I haven't figure out how LightGBM Dask finds the bin boundaries yet. But given that the CLI distributed version uses only local data for bin boundary finding when `pre_partition=True`, then a difference in bin boundaries between CLI and Dask is expected as long as Dask does not use the same partition.

Could you please point me the code where I can find how Dask partitions the data? Thanks!

**Comment by jmoralez** [comment]
The partitions are identical, since I create them manually in the script. What I meant is that even with the CLI version alone, the results differ depending on the order of the machines. So in this case we have two machines A, B and two partitions P1 and P2. If A gets P1 and B gets P2 then the threshold is `0.08647`, however if machine A gets P2 and machine B gets P1 then the threshold is `0.09008`.

Here's a smaller example:
<details>
  <summary>gen_data.py</summary>

  ```python
import numpy as np
from sklearn.datasets import make_regression


X, y = make_regression(n_samples=1_000, n_features=4, n_informative=2, random_state=42)
data = np.hstack([y.reshape(-1, 1), X])
partitions = np.array_split(data, 2)
for i, partition in enumerate(partitions):
    np.savetxt(f'train{i}.txt', partition, delimiter=',')
  ```
</details>
<details>
  <summary>train0.conf</summary>

  ```
output_model = model0.txt
local_listen_port = 49524
data = train0.txt
task = train
pre_partition = True
machine_list_file = mlist.txt
tree_learner = data
force_row_wise = True
verbose = 0
num_boost_round = 1
num_leaves = 3
num_threads = 2
objective = regression
num_machines = 2
  ```
</details>
<details>
  <summary>train1.conf</summary>

  ```
output_model = model1.txt
local_listen_port = 49525
data = train1.txt
task = train
pre_partition = True
machine_list_file = mlist.txt
tree_learner = data
force_row_wise = True
verbose = 0
num_boost_round = 1
num_leaves = 3
num_threads = 2
objective = regression
num_machines = 2
  ```
</details>
<details>
  <summary>mlist.txt</summary>

  ```
127.0.0.1 49524
127.0.0.1 49525
  ```
</details>

If you run this with `train0.conf` having the first machine in `mlist.txt` (as it is right now) you get `0.08647` in the first split. However if you reverse the order in `mlist.txt` such that `train0.conf` has the second machine, the split now becomes `0.09008`. That's what I meant when I asked if this behavior was expected, hope this helps @shiyu1994.

**Comment by jameslamb** [comment]
@jmoralez, @shiyu1994 's comment gave me an idea. I think we should always set  `pre_partition=True` in the Dask interface.

https://github.com/microsoft/LightGBM/blob/dac0dffeb2837934647b374eab8469f91fc6a214/include/LightGBM/config.h#L641-L644

I think that since that interface assumes that Dask is handling data distribution (through the way your Dask DataFrame or Dask Array is partitioned), from LightGBM's perspective `pre_partition` should always be true.

What do you think?

To be clear, I don't think such a change would change what you observed in https://github.com/microsoft/LightGBM/issues/3835#issuecomment-961999914, since the config value is only used when loading a dataset  from a file (see these links I found with `git grep pre_partition`).

`DatasetLoader::LoadFromFile()`

https://github.com/microsoft/LightGBM/blob/aab212a78285fb245cf980d39e62a1048af5005f/src/io/dataset_loader.cpp#L184

`DatasetLoader::LoadFromBinFile()`

https://github.com/microsoft/LightGBM/blob/aab212a78285fb245cf980d39e62a1048af5005f/src/io/dataset_loader.cpp#L520

`DatasetLoader::LoadTextDataToMemory()`

https://github.com/microsoft/LightGBM/blob/aab212a78285fb245cf980d39e62a1048af5005f/src/io/dataset_loader.cpp#L863

`DatasetLoader::SampleTextDataFromFile()`

https://github.com/microsoft/LightGBM/blob/aab212a78285fb245cf980d39e62a1048af5005f/src/io/dataset_loader.cpp#L926

**Comment by shiyu1994** [comment]
@jmoralez When using data distributed training, bin finding of different features are distributed to the machines by **feature**. So in your example, machine with rank 0 will always finds the bin mappers for the first part of the features, and the machine with rank 1 will always finds the bin mappers for the second part of the features. Both with their **local** data. So I think the difference make sense, because with different rank numbers, different data will be used to find the bin mappers of the same feature.

**Comment by jmoralez** [comment]
> I think that since that interface assumes that Dask is handling data distribution (through the way your Dask DataFrame or Dask Array is partitioned), from LightGBM's perspective pre_partition should always be true.
What do you think?

@jameslamb I agree. Although we should probably not let the user change this so maybe we should add this in `_train`? https://github.com/microsoft/LightGBM/blob/b1facf50502fb51a1e60c05b3ec83f68289df497/python-package/lightgbm/dask.py#L388

Thank you for the explanation @shiyu1994. I think that may be the cause of the differences we see in the distributed tests. Will report back soon after I run some tests.

**Comment by jmoralez** [comment]
I've been running some more experiments and it definitely seems that the differences we sometimes observe in the predictions between local and distributed are due to the slightly different thresholds, because when a sample is very close to the thresholds it can end up in the opposite direction of a split with respect to the other model.

The most dramatic case is with two leaves only, where even though the trees are almost identical, a sample gets a very different value.
<details>
  <summary>Sample code</summary>

  ```python
  import dask.dataframe as dd
  import lightgbm as lgb
  import matplotlib.pyplot as plt
  import numpy as np
  import pandas as pd
  from dask.distributed import Client
  from sklearn.datasets import make_regression
  
  
  client = Client(n_workers=2, threads_per_worker=1)
  X, y = make_regression(n_samples=1_000, n_features=4, n_informative=2, random_state=42)
  X = pd.DataFrame(X, columns=[f'feature_{i}' for i in range(X.shape[1])])
  y = pd.Series(y)
  dX = dd.from_pandas(X, npartitions=2)
  dy = dd.from_pandas(y, npartitions=2)
  params = {
      "boosting_type": "gbdt",
      "random_state": 42,
      "num_leaves": 2,
      "n_estimators": 20,
      "tree_learner": "data",
  }
  local_reg = lgb.LGBMRegressor(**params)
  local_reg.fit(X, y)
  local_preds = local_reg.predict(X)
  
  dask_reg = lgb.DaskLGBMRegressor(**params)
  dask_reg.fit(dX, dy)
  dask_preds = dask_reg.predict(dX).compute()
  
  pct_diffs = abs(local_preds - dask_preds) / local_preds
  top_diff_idx = np.argsort(-pct_diffs)[0]
  x = X.iloc[top_diff_idx]
  
  fig, ax = plt.subplots(ncols=2, figsize=(10, 6))
  
  for reg, title, axi in zip([local_reg, dask_reg], ['local', 'distributed'], ax.flat):
      lgb.plot_tree(reg, ax=axi, show_info=['internal_value', 'leaf_count'], tree_index=0, precision=5)
      axi.set_title(title);

  ```
</details>

![simple](https://user-images.githubusercontent.com/8473587/146664029-c2b3f55c-1f36-406f-9fd4-81c90b7c8658.png)

With a higher number of leaves the differences are smaller but they increase as the iterations go by, for example using 8 leaves with the code above the prediction per iteration for a specific sample looks like this:

<details>
  <summary>Additional code for plot</summary>

  ```python
local_result = []
dask_result = []
dask_bst = dask_reg.booster_

for i in range(1, 21):
    local_result.append(local_regressor.predict(x, num_iteration=i)[0])
    dask_result.append(dask_bst.predict(x, num_iteration=i)[0])
    
fig, ax = plt.subplots(figsize=(6, 4))
pd.DataFrame({'local': local_result, 'dask': dask_result}).plot(marker='.', ax=ax)
  ```
</details>

![per_iteration](https://user-images.githubusercontent.com/8473587/146664182-6e6990da-c858-4aa8-810f-2b5e2232fd2d.png)
 
From the plot above it can be seen that at tree index 6 the predictions become very different. The structure of the built trees and the path the sample takes through them can be seen in the image below, as well as the predicted value.

![trees](https://user-images.githubusercontent.com/8473587/146663984-5df41113-3a40-487f-bf51-4ef8e93a5ee5.png)

We can see here that the slight differences seem to accumulate and by tree index 5 the tree structures start to differ.

**Comment by jameslamb** [comment]
Nice investigation @jmoralez !

I think this still makes sense given @shiyu1994 's explanation in https://github.com/microsoft/LightGBM/issues/3835#issuecomment-962758221. Since the Dask code is splitting `X` in half, and it was already not very big (1000 rows), it makes sense to me that the bin boundaries determined based on 500 samples might be slightly different than those base on 1000 samples.

To eliminate that specific source of difference and check for other types (like maybe loss of precision during collective operations like syncing histograms), you could try create a Dask DataFrame with two partitions where each partition is a full copy of `X`. Doing that, I'd expect the histogram boundaries found to be identical to those found from local training.

**Comment by shiyu1994** [comment]
@jmoralez Nice investigation. I think the only way to eliminate the difference is by synchronizing the processes when finding the bin boundaries. But that would require designing a distributed algorithm for bin finding, which is nontrivial.
